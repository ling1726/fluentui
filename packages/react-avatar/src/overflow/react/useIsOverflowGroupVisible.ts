import { useOverflowContext } from './overflowContext';

export function useIsOverflowGroupVisible(id: string | number) {
  return useOverflowContext(ctx => ctx.groupVisibility[id + '']);
}
