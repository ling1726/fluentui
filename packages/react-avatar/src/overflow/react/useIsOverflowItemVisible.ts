import { useOverflowContext } from './overflowContext';

export function useIsOverflowItemVisible(id: string | number) {
  return !!useOverflowContext(ctx => ctx.itemVisibility[id]);
}
