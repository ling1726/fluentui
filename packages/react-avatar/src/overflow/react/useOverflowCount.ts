import { useOverflowContext } from './overflowContext';

export const useOverflowCount = () =>
  useOverflowContext(v => {
    return Object.entries(v.itemVisibility).reduce((acc, cur) => {
      if (!cur[0].startsWith('divider') && !cur[1]) {
        acc++;
      }

      return acc;
    }, 0);
  });
