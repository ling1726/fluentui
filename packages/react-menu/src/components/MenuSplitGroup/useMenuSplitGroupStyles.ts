import { makeStyles, mergeClasses } from '@fluentui/react-make-styles';
import type { MenuSplitGroupState } from './MenuSplitGroup.types';

/**
 * Styles for the root slot
 */
const useStyles = makeStyles({
  root: theme => ({
    display: 'flex',
    '& > [role="menuitem"]:nth-child(1)': {
      width: '100%',
    },
    '& > [role="menuitem"]:nth-child(2)': {
      borderTopLeftRadius: 0,
      borderBottomLeftRadius: 0,
      paddingLeft: 0,
      marginLeft: 'auto',
      ':before': {
        content: '""',
        width: theme.strokeWidthThin,
        height: '24px',
        backgroundColor: theme.colorNeutralStroke1,
      },
    },
  }),
});

/**
 * Apply styling to the MenuSplitGroup slots based on the state
 */
export const unstable_useMenuSplitGroupStyles = (state: MenuSplitGroupState): MenuSplitGroupState => {
  const styles = useStyles();
  state.root.className = mergeClasses(styles.root, state.root.className);
  return state;
};
