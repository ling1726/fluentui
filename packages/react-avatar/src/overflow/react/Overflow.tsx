import * as React from 'react';
import { useMergedRefs } from '@fluentui/react-utilities';
import { OverflowGroupState, OnUpdateOverflow, OverflowEventPayload } from '../native/overflowManager';
import { OverflowContext } from './overflowContext';
import { useOverflowContainer, UseOverflowContainerOptions } from './useOverflowContainer';

export type OverflowChildProps = React.HTMLAttributes<HTMLElement> & {
  ref?: React.Ref<never>;
};

export type OverflowProps = UseOverflowContainerOptions & {
  children: React.ReactElement;
};

export const Overflow = React.forwardRef((props: OverflowProps, ref) => {
  const [hasOverflow, setHasOverflow] = React.useState(false);
  const [itemVisibility, setItemVisibility] = React.useState<Record<string, boolean>>({});
  const [groupVisibility, setGroupVisibility] = React.useState<Record<string, OverflowGroupState>>({});

  const updateItemVisibility: OnUpdateOverflow = (data: OverflowEventPayload) => {
    setHasOverflow(() => data.invisibleItems.length > 0);
    setItemVisibility(() => {
      const newState: Record<string, boolean> = {};
      data.visibleItems.forEach(x => (newState[x.id] = true));
      data.invisibleItems.forEach(x => (newState[x.id] = false));
      return newState;
    });
    setGroupVisibility(groupVisibility);
  };

  const { containerRef, registerItem, updateOverflow } = useOverflowContainer(updateItemVisibility, {
    ...props,
  });

  const mergedRef = useMergedRefs(containerRef, ref);

  const children = React.cloneElement(React.Children.only(props.children), { ref: mergedRef });

  return (
    <OverflowContext.Provider
      value={{
        itemVisibility,
        groupVisibility,
        hasOverflow,
        registerItem,
        updateOverflow,
      }}
    >
      {children}
    </OverflowContext.Provider>
  );
});
