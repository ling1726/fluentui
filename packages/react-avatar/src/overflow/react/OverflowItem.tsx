import * as React from 'react';
import { useMergedRefs } from '@fluentui/react-utilities';
import { useOverflowItem } from './useOverflowItem';

export type OverflowItemChildProps = React.HTMLAttributes<HTMLElement> & {
  ref?: React.Ref<never>;
};

export type OverflowItemProps = {
  id: string | number;
  groupId?: string | number;
  priority?: number;
  children: React.ReactElement;
};

export const OverflowItem = React.forwardRef((props: OverflowItemProps, ref) => {
  const { id, groupId, priority, children } = props;

  const containerRef = useOverflowItem<HTMLDivElement>(id, priority, groupId);

  const mergedRef = useMergedRefs(containerRef, ref);

  return React.cloneElement(React.Children.only(children), { ref: mergedRef, id, groupId, priority });
});
