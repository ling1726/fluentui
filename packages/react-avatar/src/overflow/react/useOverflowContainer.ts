import * as React from 'react';
import { useEventCallback } from '@fluentui/react-utilities';
import {
  ObserveOptions,
  OnUpdateItemVisibility,
  OnUpdateOverflow,
  OverflowItemEntry,
  OverflowManager,
} from '../native/overflowManager';

export interface UseOverflowContainerOptions extends ObserveOptions {}

export const useOverflowContainer = (update: OnUpdateOverflow, options: UseOverflowContainerOptions = {}) => {
  const { overflowAxis, overflowDirection, padding, minimumVisible, onUpdateItemVisibility } = options;
  // DOM ref to the overflow container element
  const containerRef = React.useRef<HTMLDivElement>(null);
  const updateOverflowItems = useEventCallback(update);
  const [overflowManager] = React.useState<OverflowManager>(() => new OverflowManager(updateOverflowItems));

  // eslint-disable-next-line no-restricted-properties
  React.useLayoutEffect(() => {
    if (!containerRef.current) {
      return;
    }

    overflowManager.observe(containerRef.current, {
      overflowDirection,
      overflowAxis,
      padding,
      minimumVisible,
      onUpdateItemVisibility: onUpdateItemVisibility ?? defaultUpdateVisibilityCallback,
    });

    return () => {
      overflowManager.disconnect();
    };
  }, [
    updateOverflowItems,
    overflowManager,
    overflowDirection,
    overflowAxis,
    padding,
    minimumVisible,
    onUpdateItemVisibility,
  ]);

  const registerItem = React.useCallback(
    (item: OverflowItemEntry) => {
      overflowManager.addItems(item);

      return () => overflowManager.removeItem(item.id);
    },
    [overflowManager],
  );

  const updateOverflow = React.useCallback(() => {
    overflowManager.updateOverflow();
  }, [overflowManager]);

  return {
    containerRef,
    registerItem,
    updateOverflow,
  };
};

const defaultUpdateVisibilityCallback: OnUpdateItemVisibility = ({ item, visible }) => {
  if (visible) {
    item.element.style.removeProperty('display');
  } else {
    item.element.style.setProperty('display', 'none');
  }
};
