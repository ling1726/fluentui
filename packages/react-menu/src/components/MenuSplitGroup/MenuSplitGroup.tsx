import * as React from 'react';
import { unstable_useMenuSplitGroup } from './useMenuSplitGroup';
import { unstable_renderMenuSplitGroup } from './renderMenuSplitGroup';
import { unstable_useMenuSplitGroupStyles } from './useMenuSplitGroupStyles';
import type { MenuSplitGroupProps } from './MenuSplitGroup.types';
import type { ForwardRefComponent } from '@fluentui/react-utilities';

/**
 * MenuSplitGroup component
 */
export const MenuSplitGroup_unstable: ForwardRefComponent<MenuSplitGroupProps> = React.forwardRef((props, ref) => {
  const state = unstable_useMenuSplitGroup(props, ref);

  unstable_useMenuSplitGroupStyles(state);
  return unstable_renderMenuSplitGroup(state);
});

MenuSplitGroup_unstable.displayName = 'MenuSplitGroup';
