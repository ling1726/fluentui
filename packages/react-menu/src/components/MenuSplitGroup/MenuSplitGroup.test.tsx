import * as React from 'react';
import { render } from '@testing-library/react';
import { MenuSplitGroup_unstable } from './MenuSplitGroup';
import { isConformant } from '../../common/isConformant';

describe('MenuSplitGroup', () => {
  isConformant({
    Component: MenuSplitGroup_unstable,
    disabledTests: ['exports-component', 'has-top-level-file', 'exported-top-level'],
    displayName: 'MenuSplitGroup',
  });

  it('renders a default state', () => {
    const result = render(<MenuSplitGroup_unstable>Default MenuSplitGroup</MenuSplitGroup_unstable>);
    expect(result.container).toMatchSnapshot();
  });
});
