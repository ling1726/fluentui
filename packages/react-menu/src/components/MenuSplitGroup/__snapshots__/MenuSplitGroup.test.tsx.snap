// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MenuSplitGroup renders a default state 1`] = `
<div>
  <div
    class="fui-MenuSplitGroup"
    role="presentation"
  >
    Default MenuSplitGroup
  </div>
</div>
`;
